File: 2023/day07/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Card strength orders (higher index = stronger)
const cardStrength = "23456789TJQKA"
const cardStrengthJoker = "J23456789TQKA" // J is weakest in Part 2

// Hand type constants (higher = stronger)
const (
	HighCard     = iota // 0 - [1,1,1,1,1]
	OnePair             // 1 - [2,1,1,1]
	TwoPair             // 2 - [2,2,1]
	ThreeOfKind         // 3 - [3,1,1]
	FullHouse           // 4 - [3,2]
	FourOfKind          // 5 - [4,1]
	FiveOfKind          // 6 - [5]
)

// Hand represents a Camel Cards hand with its cards and bid amount.
type Hand struct {
	cards string
	bid   int
}

// classifyFromCounts determines hand type from sorted card counts (descending).
func classifyFromCounts(values []int) int {
	switch {
	case len(values) == 1 || values[0] == 5:
		return FiveOfKind
	case values[0] == 4:
		return FourOfKind
	case values[0] == 3 && len(values) > 1 && values[1] == 2:
		return FullHouse
	case values[0] == 3:
		return ThreeOfKind
	case values[0] == 2 && len(values) > 1 && values[1] == 2:
		return TwoPair
	case values[0] == 2:
		return OnePair
	default:
		return HighCard
	}
}

// getHandType returns the hand type as an integer (higher = stronger).
func getHandType(hand string) int {
	counts := make(map[rune]int)
	for _, c := range hand {
		counts[c]++
	}

	// Get sorted counts (descending)
	var values []int
	for _, v := range counts {
		values = append(values, v)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(values)))

	return classifyFromCounts(values)
}

// getHandTypeWithJokers returns hand type with J as wildcards.
func getHandTypeWithJokers(hand string) int {
	jokerCount := strings.Count(hand, "J")
	if jokerCount == 0 {
		return getHandType(hand)
	}
	if jokerCount == 5 {
		return FiveOfKind
	}

	// Count non-joker cards
	counts := make(map[rune]int)
	for _, c := range hand {
		if c != 'J' {
			counts[c]++
		}
	}

	// Get sorted counts (descending)
	var values []int
	for _, v := range counts {
		values = append(values, v)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(values)))

	// Add jokers to the highest count (optimal strategy)
	values[0] += jokerCount

	return classifyFromCounts(values)
}

// handKey returns a sort key for comparing hands
func handKey(hand string, strengthOrder string) []int {
	key := make([]int, len(hand))
	for i, c := range hand {
		key[i] = strings.IndexRune(strengthOrder, c)
	}
	return key
}

// compareHands compares two hands for sorting (returns true if a < b)
func compareHands(a, b Hand, useJokers bool) bool {
	var typeA, typeB int
	var strengthOrder string

	if useJokers {
		typeA = getHandTypeWithJokers(a.cards)
		typeB = getHandTypeWithJokers(b.cards)
		strengthOrder = cardStrengthJoker
	} else {
		typeA = getHandType(a.cards)
		typeB = getHandType(b.cards)
		strengthOrder = cardStrength
	}

	if typeA != typeB {
		return typeA < typeB
	}

	// Compare card by card
	keyA := handKey(a.cards, strengthOrder)
	keyB := handKey(b.cards, strengthOrder)
	for i := range keyA {
		if keyA[i] != keyB[i] {
			return keyA[i] < keyB[i]
		}
	}
	return false
}

func parseInput(filename string) ([]Hand, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var hands []Hand
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Fields(line)
		bid, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("invalid bid value %q: %w", parts[1], err)
		}
		hands = append(hands, Hand{cards: parts[0], bid: bid})
	}
	return hands, scanner.Err()
}

func part1(hands []Hand) int {
	// Make a copy to avoid modifying original
	sorted := make([]Hand, len(hands))
	copy(sorted, hands)

	sort.Slice(sorted, func(i, j int) bool {
		return compareHands(sorted[i], sorted[j], false)
	})

	total := 0
	for rank, hand := range sorted {
		total += (rank + 1) * hand.bid
	}
	return total
}

func part2(hands []Hand) int {
	// Make a copy to avoid modifying original
	sorted := make([]Hand, len(hands))
	copy(sorted, hands)

	sort.Slice(sorted, func(i, j int) bool {
		return compareHands(sorted[i], sorted[j], true)
	})

	total := 0
	for rank, hand := range sorted {
		total += (rank + 1) * hand.bid
	}
	return total
}

func main() {
	hands, err := parseInput("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(hands))
	fmt.Printf("Part 2: %d\n", part2(hands))
}

File: 2023/day07/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.IntStream;

public class Solution {

    // Hand type constants (higher = stronger)
    private static final int HIGH_CARD = 0;
    private static final int ONE_PAIR = 1;
    private static final int TWO_PAIR = 2;
    private static final int THREE_OF_A_KIND = 3;
    private static final int FULL_HOUSE = 4;
    private static final int FOUR_OF_A_KIND = 5;
    private static final int FIVE_OF_A_KIND = 6;

    // Card strength order (higher index = stronger)
    private static final String CARD_STRENGTH = "23456789TJQKA";
    private static final String CARD_STRENGTH_JOKER = "J23456789TQKA";

    public static void main(String[] args) throws IOException {
        var input = Files.readString(Path.of("../input.txt")).trim();
        var lines = input.split("\n");

        System.out.println("Part 1: " + part1(lines));
        System.out.println("Part 2: " + part2(lines));
    }

    private static long part1(String[] lines) {
        var hands = Arrays.stream(lines)
                .map(line -> {
                    var parts = line.split("\\s+");
                    return new Hand(parts[0], Integer.parseInt(parts[1]));
                })
                .sorted(Comparator.comparingInt((Hand h) -> getHandType(h.cards))
                        .thenComparing(h -> getCardValues(h.cards, CARD_STRENGTH)))
                .toList();

        return IntStream.rangeClosed(1, hands.size())
                .mapToLong(rank -> (long) rank * hands.get(rank - 1).bid)
                .sum();
    }

    private static long part2(String[] lines) {
        var hands = Arrays.stream(lines)
                .map(line -> {
                    var parts = line.split("\\s+");
                    return new Hand(parts[0], Integer.parseInt(parts[1]));
                })
                .sorted(Comparator.comparingInt((Hand h) -> getHandTypeWithJokers(h.cards))
                        .thenComparing(h -> getCardValues(h.cards, CARD_STRENGTH_JOKER)))
                .toList();

        return IntStream.rangeClosed(1, hands.size())
                .mapToLong(rank -> (long) rank * hands.get(rank - 1).bid)
                .sum();
    }

    private static int getHandType(String hand) {
        var counts = getCardCounts(hand);
        Arrays.sort(counts);
        return classifyHand(counts);
    }

    private static int getHandTypeWithJokers(String hand) {
        var jokerCount = (int) hand.chars().filter(c -> c == 'J').count();

        if (jokerCount == 0) {
            return getHandType(hand);
        }
        if (jokerCount == 5) {
            return FIVE_OF_A_KIND;
        }

        // Count non-joker cards
        var counts = new int[13];
        for (char c : hand.toCharArray()) {
            if (c != 'J') {
                var idx = CARD_STRENGTH.indexOf(c);
                counts[idx]++;
            }
        }

        Arrays.sort(counts);
        // Add jokers to the highest count
        counts[12] += jokerCount;

        return classifyHand(counts);
    }

    private static int[] getCardCounts(String hand) {
        var counts = new int[13];
        for (char c : hand.toCharArray()) {
            var idx = CARD_STRENGTH.indexOf(c);
            counts[idx]++;
        }
        Arrays.sort(counts);
        return counts;
    }

    private static int classifyHand(int[] counts) {
        // counts is sorted ascending, so counts[12] is highest, counts[11] is second highest
        var highest = counts[12];
        var secondHighest = counts[11];

        if (highest == 5) return FIVE_OF_A_KIND;
        if (highest == 4) return FOUR_OF_A_KIND;
        if (highest == 3 && secondHighest == 2) return FULL_HOUSE;
        if (highest == 3) return THREE_OF_A_KIND;
        if (highest == 2 && secondHighest == 2) return TWO_PAIR;
        if (highest == 2) return ONE_PAIR;
        return HIGH_CARD;
    }

    private static CardValues getCardValues(String hand, String cardStrength) {
        var values = new int[5];
        for (var i = 0; i < 5; i++) {
            values[i] = cardStrength.indexOf(hand.charAt(i));
        }
        return new CardValues(values);
    }

    private record Hand(String cards, int bid) {}

    private record CardValues(int[] values) implements Comparable<CardValues> {
        @Override
        public int compareTo(CardValues other) {
            return Arrays.compare(this.values, other.values);
        }
    }
}
